packages/rpc/common/src/ts/websocket/WebSocketServerStreamStub.ts

```
@@ -1,72 +1,19 @@
-import {CompleteListener, MessageListener, parse, send} from './utils';
-import {COMPLETE_METHOD} from '../stub/reservedRpcMethods';
-import {has} from '@0cfg/utils-common/lib/has';
-import {execAll} from '@0cfg/utils-common/lib/execAll';
+import {WebSocketStreamStub} from './WebSocketStreamStub';
 
 export class WebSocketServerStreamStub<ClientMessageT, ServerMessageT>
+    extends WebSocketStreamStub<ClientMessageT, ServerMessageT>
     implements ServerStreamStub<ClientMessageT, ServerMessageT> {
 
-    private readonly socket: CommonReconnectingWebSocket;
-    private requestId: number | undefined;
-    private readonly method: string;
-    private readonly messageListeners: MessageListener<ServerMessageT>[] = [];
-    private readonly completeListeners: CompleteListener[] = [];
-    private readonly sequential: Sequential;
-
     public constructor(socket: CommonReconnectingWebSocket, requestIdSequential: Sequential, method: string) {
-        this.socket = socket;
-        this.sequential = requestIdSequential;
-        this.method = method;
-        this.socket.onClose(message =>
-            this.completeListeners.forEach(listener =>
-                listener(errStatus(message))));
-        this.socket.onMessage(data => this.parseAndForwardServerMessage(data));
-    }
-
-    public complete(end: Reply): void {
-        send<SerializedReply<never>>(this.socket, {
-            requestId: this.requestId!,
-            method: COMPLETE_METHOD,
-            args: end.toSerializedReply(),
-        });
-        this.requestId = undefined;
-    }
-
-    public onCompleted(listener: CompleteListener): void {
-        this.completeListeners.push(listener);
+        super(socket, requestIdSequential, method);
     }
 
-    public onMessage(listener: MessageListener<ServerMessageT>): void {
-        this.messageListeners.push(listener);
+    public start(message: ClientMessageT): void {
+        super.send(message);
+        this.complete(getOk());
```

Why would you complete right after starting?

---

Don't know. This is master behaviour.

---

Master completes if there was already a requestId (if a running or completed stream is restarted).
In a previous comment we agreed that a stream should never be restarted but recreated by the stream factory.
The complete here can go away but we should throw an exception if we are already running or have been completed.

---

Ok, removed the line and added a check.